File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/stored/sync/StoredFileSynchronization.java
package com.lasthopesoftware.bluewater.client.stored.sync;

import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.v4.content.LocalBroadcastManager;
import com.annimon.stream.Stream;
import com.lasthopesoftware.bluewater.client.connection.builder.BuildUrlProviders;
import com.lasthopesoftware.bluewater.client.library.access.ILibraryProvider;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.exceptions.StoredFileJobException;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.exceptions.StoredFileReadException;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.exceptions.StoredFileWriteException;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.repository.StoredFile;
import com.lasthopesoftware.bluewater.client.stored.library.sync.LibrarySyncHandler;
import com.lasthopesoftware.bluewater.client.stored.library.sync.factory.ProduceLibrarySyncHandlers;
import com.lasthopesoftware.bluewater.shared.MagicPropertyBuilder;
import com.lasthopesoftware.bluewater.shared.observables.ObservedPromise;
import com.lasthopesoftware.bluewater.shared.observables.StreamedPromise;
import com.lasthopesoftware.storage.write.exceptions.StorageCreatePathException;
import io.reactivex.Completable;
import io.reactivex.exceptions.CompositeException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class StoredFileSynchronization implements SynchronizeStoredFiles {

	private static final MagicPropertyBuilder magicPropertyBuilder = new MagicPropertyBuilder(StoredFileSynchronization.class);

	public static final String onSyncStartEvent = magicPropertyBuilder.buildProperty("onSyncStartEvent");
	public static final String onSyncStopEvent = magicPropertyBuilder.buildProperty("onSyncStopEvent");
	public static final String onFileQueuedEvent = magicPropertyBuilder.buildProperty("onFileQueuedEvent");
	public static final String onFileDownloadingEvent = magicPropertyBuilder.buildProperty("onFileDownloadingEvent");
	public static final String onFileDownloadedEvent = magicPropertyBuilder.buildProperty("onFileDownloadedEvent");
	public static final String onFileWriteErrorEvent = magicPropertyBuilder.buildProperty("onFileWriteErrorEvent");
	public static final String onFileReadErrorEvent = magicPropertyBuilder.buildProperty("onFileReadErrorEvent");
	public static final String storedFileEventKey = magicPropertyBuilder.buildProperty("storedFileEventKey");

	private static final Logger logger = LoggerFactory.getLogger(StoredFileSynchronization.class);
	private final ILibraryProvider libraryProvider;
	private final LocalBroadcastManager localBroadcastManager;
	private final BuildUrlProviders urlProviders;
	private final ProduceLibrarySyncHandlers librarySyncHandlersProduction;

	public StoredFileSynchronization(
		ILibraryProvider libraryProvider,
		LocalBroadcastManager localBroadcastManager,
		BuildUrlProviders urlProviders,
		ProduceLibrarySyncHandlers librarySyncHandlersProduction) {
		this.libraryProvider = libraryProvider;
		this.localBroadcastManager = localBroadcastManager;
		this.urlProviders = urlProviders;
		this.librarySyncHandlersProduction = librarySyncHandlersProduction;
	}

	@Override
	public Completable streamFileSynchronization() {
		logger.info("Starting sync.");

		localBroadcastManager.sendBroadcast(new Intent(onSyncStartEvent));

		return StreamedPromise.stream(libraryProvider.getAllLibraries())
			.map(library -> {
				if (library.isSyncLocalConnectionsOnly()) library.setLocalOnly(true);

				return ObservedPromise.observe(urlProviders.promiseBuiltUrlProvider(library)
					.then(urlProvider -> librarySyncHandlersProduction.getNewSyncHandler(urlProvider, library)));
			})
			.flatMap(o -> o.flatMap(LibrarySyncHandler::observeLibrarySync, true))
			.flatMapCompletable(storedFileJobStatus -> {
				switch (storedFileJobStatus.storedFileJobState) {
					case Queued:
						sendStoredFileBroadcast(onFileQueuedEvent, storedFileJobStatus.storedFile);
						return Completable.complete();
					case Downloading:
						sendStoredFileBroadcast(onFileDownloadingEvent, storedFileJobStatus.storedFile);
						return Completable.complete();
					case Downloaded:
						sendStoredFileBroadcast(onFileDownloadedEvent, storedFileJobStatus.storedFile);
						return Completable.complete();
				}

				return Completable.complete();
			}, true)
			.onErrorComplete(this::handleError)
			.doOnComplete(this::sendStoppedSync)
			.doOnDispose(this::sendStoppedSync);
	}

	private boolean handleError(Throwable e) {
		if (e instanceof CompositeException) {
			final CompositeException compositeException = (CompositeException)e;
			return Stream.of(compositeException.getExceptions()).allMatch(this::handleError);
		}

		if (e instanceof StoredFileWriteException) {
			sendStoredFileBroadcast(onFileWriteErrorEvent, ((StoredFileWriteException)e).getStoredFile());
			return true;
		}

		if (e instanceof StoredFileReadException) {
			sendStoredFileBroadcast(onFileReadErrorEvent, ((StoredFileReadException)e).getStoredFile());
			return true;
		}

		return e instanceof StorageCreatePathException || e instanceof StoredFileJobException;
	}

	private void sendStoppedSync() {
		localBroadcastManager.sendBroadcast(new Intent(onSyncStopEvent));
	}

	private void sendStoredFileBroadcast(@NonNull String action, @NonNull StoredFile storedFile) {
		final Intent storedFileBroadcastIntent = new Intent(action);
		storedFileBroadcastIntent.putExtra(storedFileEventKey, storedFile.getId());
		localBroadcastManager.sendBroadcast(storedFileBroadcastIntent);
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/stored/sync/StoredFileSynchronization.kt
package com.lasthopesoftware.bluewater.client.stored.sync

import com.lasthopesoftware.bluewater.client.browsing.library.access.ProvideLibraries
import com.lasthopesoftware.bluewater.client.browsing.library.repository.libraryId
import com.lasthopesoftware.bluewater.client.stored.library.items.files.PruneStoredFiles
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.StoredFileJobState
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.exceptions.StoredFileJobException
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.exceptions.StoredFileReadException
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.exceptions.StoredFileWriteException
import com.lasthopesoftware.bluewater.client.stored.library.sync.CheckForSync
import com.lasthopesoftware.bluewater.client.stored.library.sync.ControlLibrarySyncs
import com.lasthopesoftware.bluewater.shared.lazyLogger
import com.lasthopesoftware.bluewater.shared.messages.application.SendApplicationMessages
import com.lasthopesoftware.bluewater.shared.observables.stream
import com.lasthopesoftware.storage.write.exceptions.StorageCreatePathException
import com.namehillsoftware.handoff.promises.Promise
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.exceptions.CompositeException

class StoredFileSynchronization(
    private val libraryProvider: ProvideLibraries,
    private val applicationMessages: SendApplicationMessages,
    private val pruneStoredFiles: PruneStoredFiles,
    private val checkSync: CheckForSync,
    private val syncHandler: ControlLibrarySyncs
) : SynchronizeStoredFiles {

	companion object {
		private val logger by lazyLogger<StoredFileSynchronization>()
	}

	override fun streamFileSynchronization(): Completable {
		logger.info("Starting sync.")
		applicationMessages.sendMessage(SyncStateMessage.SyncStarted)
		return Promise.Proxy { cp ->
				pruneStoredFiles
					.pruneDanglingFiles()
					.also(cp::doCancel)
					.eventually { checkSync.promiseIsSyncNeeded() }
					.eventually { isNeeded ->
						if (isNeeded && !cp.isCancelled) libraryProvider.promiseAllLibraries()
						else Promise(emptyList())
					}
			}
			.stream()
			.flatMap({ library -> syncHandler.observeLibrarySync(library.libraryId) }, true)
			.flatMapCompletable({ storedFileJobStatus ->
				val message = when (storedFileJobStatus.storedFileJobState) {
					StoredFileJobState.Queued -> StoredFileMessage.FileQueued(storedFileJobStatus.storedFile.id)
					StoredFileJobState.Downloading -> StoredFileMessage.FileDownloading(storedFileJobStatus.storedFile.id)
					StoredFileJobState.Downloaded -> StoredFileMessage.FileDownloaded(storedFileJobStatus.storedFile.id)
					else ->	null
				}

				message?.let(applicationMessages::sendMessage)
				Completable.complete()
			}, true)
			.onErrorComplete(::handleError)
			.doOnComplete(::sendStoppedSync)
			.doOnDispose(::sendStoppedSync)
	}

	private fun handleError(e: Throwable): Boolean {
		return when (e) {
			is CompositeException -> e.exceptions.all { handleError(it) }
			is StoredFileWriteException -> {
				applicationMessages.sendMessage(StoredFileMessage.FileWriteError(e.storedFile.id))
				return true
			}
			is StoredFileReadException -> {
				applicationMessages.sendMessage(StoredFileMessage.FileReadError(e.storedFile.id))
				return true
			}
			is StorageCreatePathException -> true
			is StoredFileJobException -> true
			else -> false
		}
	}

	private fun sendStoppedSync() {
		applicationMessages.sendMessage(SyncStateMessage.SyncStopped)
	}
}
